src/adaptations/device-layer/EFR32/Entropy.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ *
+ *    Copyright (c) 2019 Nest Labs, Inc.
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Provides implementations for the Weave entropy sourcing functions
+ *          on the Silcon Labs EFR32 platforms.
+ */
+
+#include <Weave/DeviceLayer/internal/WeaveDeviceLayerInternal.h>
+#include <Weave/Support/crypto/WeaveRNG.h>
+#include <openthread/random_crypto.h>
+#include <mbedtls/entropy_poll.h>
+
+using namespace ::nl;
+using namespace ::nl::Weave;
+
+#if !WEAVE_CONFIG_RNG_IMPLEMENTATION_NESTDRBG
+#error "Nest DRBG implementation must be enabled on EFR32 platforms"
+#endif // !WEAVE_CONFIG_RNG_IMPLEMENTATION_NESTDRBG
+
+namespace nl {
+namespace Weave {
+namespace DeviceLayer {
+namespace Internal {
+
+#define WEAVE_ENTROPY_GEN_FAILED 1
+
+extern "C" int mbedtls_hardware_poll(void *data, unsigned char *output, size_t len, size_t *olen);
+
+/**
+ * Retrieve entropy from the underlying RNG source.
+ *
+ * This function is called by the Nest DRBG to acquire entropy.
+ */
+int GetEntropy_EFR32(uint8_t *buf, size_t count)
+{
+    int    res         = 0;
+    size_t entropy_len = 0;
+    size_t olen        = 0;
+
+    VerifyOrDie(count <= UINT16_MAX);
+
+    while (entropy_len < count)
+    {
+        res = mbedtls_hardware_poll(NULL, buf + entropy_len, count - entropy_len, &olen);
```

Comment:
Entropy for BLE uses a different mechanism from the radio so will not conflict.

In OpenThread on the MG12 entropy is from the ADC0.  For OpenWeave on the MG12 I've called into the OpenThread Entropy API and protected it with the ThreadStackMgr lock.

For OpenThread and OpenWeave on the MG21 both use mbedtls_hardware_poll. This uses the Secure-Element Crypto/Entropy hardware acceleration whose access is internally is protected by a FreeRTOS mutex.

I have now enabled FreeRTOS threading support.  Initialization of mbedtls threading support now occurs in main.cpp of the lock example with a call to Threading_setup.

Reply:
Regarding the MG21, since it uses the mbedtls entropy infrastructure, which provides a DRBG implementation, I think the Nest DRBG should be turned off in this case, since it adds memory and CPU overhead that is unnecessary.  To achieve this, you would set WEAVE_CONFIG_RNG_IMPLEMENTATION_NESTDRBG to 0, and WEAVE_CONFIG_RNG_IMPLEMENTATION_PLATFORM to 1, and then define InitEntropy() and GetSecureRandomData() in Entropy.c to map to mbedtls.

All that said, I'm fine with leaving the code as-is and addressing this in a later commit.